tests/test_checkout.py

```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
+        co = repo.checkout(write=True)
+        dset = co.datasets.init_dataset('dset', prototype=array5by7)
+        co.commit()
+        co.close()
+        co = repo.checkout(write=True)
+        # TODO: what is the design here. IMO, the old instance should not execute
+        dset.add(array5by7, '1')
+        co.commit()
+        co.close()
+        # TODO: even after the close, I could still fetch the
+        #   data from the old dset object.
+        print(dset['1'])
+
+    def test_operate_on_closed_checkout(self, repo, array5by7):
+        co = repo.checkout(write=True)
+        co.datasets.init_dataset('dset', prototype=array5by7)
+        co.commit()
+        co.close()
+        # TODO: This should raise permission error or something
+        with pytest.raises(KeyError):
+            co.datasets['dset']['1'] = array5by7
+        co.metadata.add('a', 'b')
+
+
+class TestBranching:
+
+    def test_invalid_name(self, repo):
+        with pytest.raises(ValueError):
+            repo.create_branch('dummy branch')
+
+    def test_new_branch_from_new_repo(self, repo):
+        branch = repo.create_branch('testbranch')
+        # TODO: it doesnt raise value error to the user
+        # instead return false and print out the value error
+        # Ideally, it should let users create new branch
+        co = repo.checkout(write=True, branch_name=branch)
+
+    def test_merge(self, written_repo, array5by7):
+        branch = written_repo.create_branch('testbranch')
+        assert type(branch) is str
+        co = written_repo.checkout(write=True, branch_name=branch)
+        assert co._branch_name == branch
+        co.datasets['_dset']['1'] = array5by7
+        co.metadata.add('a', 'b')
+        co.commit()
+        co.close()
+        written_repo.merge('test merge', 'master', branch)
+        co = written_repo.checkout()
+        # TODO: It's good to give access to name of the branch from Readcheckout
+        # assert co._branch_name == 'master'
+        assert (co.datasets['_dset']['1'] == array5by7).all()
+        assert co.metadata.get('a') == 'b'
+
+    def test_merge_without_closing_previous_checkout(self, written_repo, array5by7):
+        branch = written_repo.create_branch('testbranch')
+        co = written_repo.checkout(write=True, branch_name=branch)
+        co.datasets['_dset']['1'] = array5by7
+        co.commit()
+        # TODO: Need better error message
+        # lockHeld check at line #235 of diff.py never happens
+        with pytest.raises(PermissionError):
+            written_repo.merge('test merge', 'master', branch)
+
+    def test_name_conflict(self, written_repo, array5by7):
+        written_repo.create_branch('testbranch')
+        with pytest.raises(ValueError):
+            written_repo.create_branch('testbranch')
+
+    def test_delete_branch(self, written_repo):
+        with pytest.raises(NotImplementedError):
+            # TODO should raise "branch doesn't exist error"
```

Comment:
I haven't figured out how we should handle removing branches yet. the logic it a bit complicated because we need to figure out what data is referenced in other commits, and which can be safely removed from disk. please create an issue and link it here for this.

Reply:
#38